README.md
# Signalk-stripcharts - Generate strip charts from Signal K live boat data and from InfluxDB past boat data

  > New in version 0.2.x: Charting past data from InfluxDB made easier (you may need to adapt your [options](#stripcharts-options-for-influxdb) )

  Note: The latest `README.md` and `README.pdf` versions are available in `GitHub`, not necessarily in `npmjs.com`, as a new npm version is not created for minor improvements to the documentation.

<div><sup><a  style="color:red;" href="#table-of-contents">Jump to TOC</a></sup></div>

## Introduction

A stripchart displays the most recent live boat data from one or more Signal K paths as a graph along a time axis (x-axis). The graph flows from left (most recent) to right: this direction allows the most recent data to be observed near the left hand side y-axis and its scale.
Legends under the x-axis identify by abbreviations what paths are charted. Here is a (strip)chart with a 10 minutes time window plotting the true wind speed (TWS) and apparent wind speed (AWS):

![simple chart](./PNG/readme1.PNG)

All charts have a y-axis with its scale and unit on the left hand side. Optionally, as shown on the chart below (right hand side), a y2-axis may also be defined with its own scale and unit. The dotted line corresponds to the plot according to the y2 axis, here the apparent wind angle (AWA). When the mouse cursor browses the chart, the values associated to the nearest data points are displayed as "tooltips".

![simple chart + y2](./PNG/readme2.PNG)

The browser window can display two such charts on top of each other. For instance you may have wind speeds on the top chart and boat speeds on the bottom chart, or wind speeds with two different time windows, say 10 minutes and 24 hours. Those two charts can be selected with a dropdown list from a set of active charts.

This documentation first explains how to chart live data from Signal K server.
Then it will explained how past boat data from InfluxDB, a time series database, can be charted as well.

Now it's time to install signalk-stripcharts and start using the default charts specifications provided.

## Table of contents

- [Signalk-stripcharts - Generate strip charts from Signal K live boat data and from InfluxDB past boat data](#signalk-stripcharts---generate-strip-charts-from-signal-k-live-boat-data-and-from-influxdb-past-boat-data)
  * [Introduction](#introduction)
  * [Table of contents](#table-of-contents)
  * [Installation](#installation)
    + [Choice 1 - Install as Webapp from Signal K Dashboard](#choice-1---install-as-webapp-from-signal-k-dashboard)
    + [Choice 2 - Install on a node server - typically the boat Signal K server](#choice-2---install-on-a-node-server---typically-the-boat-signal-k-server)
    + [Choice 3 - Install on a client device](#choice-3---install-on-a-client-device)
  * [Basic usage](#basic-usage)
    + [y- and y2-axis buttons](#y--and-y2-axis-buttons)
    + [Drop-down list](#drop-down-list)
    + [Main buttons](#main-buttons)
  * [How it works](#how-it-works)
  * [Customization](#customization)
    + [Chart specifications](#chart-specifications)
      - [Derived data](#derived-data)
      - [Filtering by sources](#filtering-by-sources)
      - [Legends and lines colors](#legends-and-lines-colors)
    + [Unit conversion](#unit-conversion)
    + [General options](#general-options)
  * [Charting past data from InfluxDB](#charting-past-data-from-influxdb)
    + [General description](#general-description)
    + [Install InfluxDB and create a database](#install-influxdb-and-create-a-database)
    + [Signalk to InfluxDB Plugin](#signalk-to-influxdb-plugin)
    + [Provide the influxDB database in the options](#provide-the-influxdb-database-in-the-options)
    + [Usage in live mode and past mode](#usage-in-live-mode-and-past-mode)
    + [Optimize with retention periods and constant queries](#optimize-with-retention-periods-and-constant-queries)
    + [Stripcharts options for InfluxDB](#stripcharts-options-for-influxdb)
  * [Another way to chart InfluxDB data: GRAPHANA](#another-way-to-chart-influxdb-data---graphana)
  * [Browser compatibility](#browser-compatibility)
  * [CPU and memory requirements on the client side](#cpu-and-memory-requirements-on-the-client-side)
  * [Troubleshooting](#troubleshooting)
  * [Limitations](#limitations)
    + [To-do functional](#to-do-functional)
    + [To-do technical](#to-do-technical)
  * [Dependencies included and licences](#dependencies-included-and-licences)
  * [Credits](#credits)

<small><i><a href='http://ecotrust-canada.github.io/markdown-toc/'>Table of contents generated with markdown-toc</a></i></small>

## Installation 

Signalk-stripcharts comes with all required dependencies (including c3 charting library, d3 visualization library and InfluxDB client library).

Choose one of the following 3 methods as applicable or desired:

### Choice 1 - Install as Webapp from Signal K Dashboard

***This is the recommended set up.***

If [Signal K node server](https://github.com/SignalK/signalk-server-node) is installed on your server:
- login to the Signal K dashboard
- open the left hand side menu
- select `Appstore`, then `Available` and find `Signalk-Stripcharts` (if it is not listed, it is probably already installed)
- click on the install icon next to the version number
- when installation is finished, click on the dashboard `Restart` button

You can now start signalk-stripcharts launcher from the dashboard `Webapps` menu.

### Choice 2 - Install on a node server - typically the boat Signal K server
- Use npm: `[sudo] npm install signalk-stripcharts`

This will install signalk-stripcharts under the closest node_modules folder higher up in the hierarchy.

### Choice 3 - Install on a client device
- Download the [ZIP file](https://github.com/edefalque/signalk-stripcharts) and unzip it in a folder.

(by default, it will be installed in a folder named `signalk-stripcharts-master`; rename the folder `signalk-stripcharts`)

By installing on a client device you minimize the server processor load. 

<div align="right"><sup><a  style="color:red;" href="#table-of-contents">↥ back to TOC</a></sup></div>

## Basic usage

Start the launch menu:
- **If installed as a Signal K Webapps**: from the browser on your client or on the Signal K server, access the Signal K dashboard (e.g. *`mysignalkserveripaddr`*`:3000/admin/#/dashboard` or `localhost:3000/admin/#/dashboard`), then start Signalk-Stripcharts from the dashboard `Webapps` menu. Or start directly with *`mysignalkserveripaddr`*`:3000/signalk-stripcharts` or `localhost:3000/signalk-stripcharts`.

You may also:
- **If installed on a (Signal K) node server**: from your client browser enter url  *`mynodeipaddr:port`*`/signalk-stripcharts`
- **If installed on a client**: double click on the file *`...path...`*`/signalk-stripcharts/index.html`

On the launch menu you can choose the Signal K server (same or distinct from the node server where signalk-stripcharts is installed) and choose a set of charts, then push the `Launch signalk-stripcharts` button. If you are connected to the Internet try first with the default selections of the launch menu.

When the charts are displayed:
- hover on a legend: the corresponding plot is highlighted, the others dimmed; the corresponding Signal K path is displayed above the legend
- hover on a plot line, a tooltip is diplayed with the values corresponding to the legends
- by clicking on a legend you can toggle the dim status of the legend and of the corresponding plot

You may wish to bookmark the launched page(s) for easier later launching. Modify the query parameters `server` & `specs` as needed, e.g. if you have defined your own charts specifications (see later section "How it works").

The following sections explain signalk-stripcharts accessing and charting only live data provided by a Signal K server: this is the default configuration at installation. A [later chapter](#charting-past-data-from-influxdb) will explain how to set up an InfluxDB database to store historical boat data and chart the past data with signalk-stripcharts.

### y- and y2-axis buttons

These buttons affect the size and position of the plotted lines corresponding to the axis:

![axis buttons](./PNG/readme3.PNG)

From left to right: 
- zoom in
- zoom out
- raise the plot lines (y-axis range is shifted)
- lower the plot lines
- try to center the plot lines at present time (x = 0)
- reset to the initial scale and position settings

These buttons appear also on top of the right hand side y2 axis if used.

### Drop-down list
You can select any chart in the set for display.
If you choose `none` the other chart will be displayed on the whole window (similar effect if the same chart is selected twice).

### Main buttons

![window buttons](./PNG/readme4.PNG)

From left to right:
- toggle pause/play plotting
- toggle day/night display
- disconnect from Signal K server (if you want to connect again: reload the page)

They apply to the whole window.

<div align="right"><sup><a  style="color:red;" href="#table-of-contents">↥ back to TOC</a></sup></div>

## How it works

A chart contents and rendition are governed by the **chart specifications**, **unit conversion** and some **general options**.

In Signal K, data types are identified as paths: e.g. the true wind speed is identified by the path `environment.wind.speedTrue`.

A chart is primarily specified by:
- a name
- the length of the time axis (`timeWindow`); typically 10 minutes, 2 hours or 24 hours
- the averaging period (`avgInterval` explained below)
- the y axis range and unit (unit conversion is taken care of)
- the same properties for the optional y2 axis
- the list of Signal K paths to be charted and corresponding abbreviated legends

For each path the charted value can be the average (`AVG`), the maximum (`MAX`) and/or the minimum (`MIN`) of the values received from Signal K during the averaging period. The averaging period (`avgInterval`) should be set such that the ratio `timeWindow`/`avgInterval` is between 300 and 1000 (i.e. a reasonable number of plots along the `timeWindow`); `avgInterval` must be longer than the sampling period defined in the Signal K subscription period (`subscribePolicy.period` option), which is typically 1 second. By default, the chart is refreshed every `avgInterval` (provided that some new data has come for that chart).
When data stops coming for a particular path, the corresponding plot line is interrupted indicating probably that the corresponding instrument was disconnected or switched off.

A chart specification is provided as a JavaScript object.
Here is the specification for the chart shown at the top of this document:

```javascript
const Wind_speeds_10min = 
    // for JavaScript variables naming rules see https://javascript.info/variables#variable-naming
    { stripChartName: "Wind_speeds_10min",   // use preferably the same as the containing object name
        timeWindow: 600,            // 10min in seconds
        avgInterval: 2,             // 2 sec
        intervalsPerRefresh: 2,     // default 1
        x: { label: "min/sec ago", tickCount: 11 },   // a tick every min (10, + 1 for zero)
        y: { unit: "Knot" },
        paths: 
        [
            { path: "environment.wind.speedTrue", AVG: "TWS" },  // average is plotted, legend is "TWS"
            { path: "environment.wind.speedApparent", AVG: "AWS" }      
        ]
    }
;
```

By default, the chart is refreshed every `avgInterval` (provided that some new data has come); however refreshing can be made slower with the `intervalsPerRefresh` property in order to spare some processing.

Related chart specification objects are grouped into a set.

When the application is started as an url, the following "query" parameters must be provided:
- `server` with the address and port of the Signal K server (defaults to the address & port the page is loaded from)
- `specs` with the file name containing the chart specifications set (without .js extension).

The application subscribes to the Signal K server deltas for all the paths in the set. The values are then continuously collected and aggregated for all charts in a set (using the streaming WebSocket API).

At any one time, two charts can be displayed as selected by the user from the drop down lists.
If the user selects `none` in one of the drop down lists, the remaining chart is extended to the whole window area.

Charts can be paused by clicking ![pause button](./PNG/readme6.PNG). When paused, data collection continues "behind the scene". And the charts display will catch up when ![play button](./PNG/readme5.PNG) is clicked. When the chart window is minimized or not in view, the charts are paused likewise and will catch up when brought to view. This minimizes processing load.

When the window is closed it is disconnected from Signal K and the buffers are "lost". In order to reconnect reload the page.
(Persistency is only provided with the InfluxDB option explained further down) 

The charts specifications sets are each stored in a .js file in the `./specs` folder.
The following specifications files are provided at installation as commented examples:
- `sail.js`
- `environment.js`
- `engines.js`
- `sources_filtering_example.js`, see [Filtering by sources](#filtering-by-sources)
- `influx_test.js`, see [Charting past data from InfluxDB](#charting-past-data-from-influxdb)

Each of them can be started in its own browser tab and run concurrently.

In a specification file, a chart specification can be derived from another chart and only the properties that differ need to be specified (e.g. a two-hours chart can be derived from a ten-minutes chart, with most of the properties inherited). Inheritance is provided at the first level of properties only; this means in particular that to add, modify or delete a path property, you must provide a new `paths` (plural) property with all its `path` (singular) properties.

<div align="right"><sup><a  style="color:red;" href="#table-of-contents">↥ back to TOC</a></sup></div>

## Customization
You will likely want to customize the chart specifications files provided at installation and listed in the previous section and create new specifications as well.
You will also want to customize:
- `units.js`, which lists units and provides conversion between Signal K units and chart units
- `stripcharts_options.js`, which provides general options governing all charts

All those files are in *`...path...`*`/signalk-stripcharts/specs/`. If installed on a Raspberry PI from the Signal K Appstore, they will be in `$HOME/.signalk/node_modules/signalk-stripcharts/specs/`.

Signalk-Stripcharts does not provide yet a way to update those files by a form interface, nor a store for them that survives installation of new Signalk-Stripcharts versions. Hence:
- copy those you want to modify in a folder outside `.../signalk-stripcharts` e.g. in `~/my-specs` where you will keep your "master copies";
- edit them or create new ones with a text editor, or better with a code editor such as Geany or Visual Studio Code;
- then copy them back to `.../signalk-stripcharts/specs/` for using them;
- after intalling a new release copy your "master copies" to `.../signalk-stripcharts/specs/` as needed.

The following subsections provides additional guidance for customizing or creating new chart specs, units and options.

### Chart specifications
The specifications files are installed in `..../signalk-stripcharts/specs/`. Ample comments are provided for those features that are not explained in this document.

The launch menu has a button that lists all paths and sources currently provided by the selected Signal K server. Switch on all your instruments and systems in order to obtain a full list of what you can chart.

#### Derived data
Some path values are provided to Signal K by the instruments; some other path values may need to be derived. E.g. true wind speed, if not provided by the instruments, can be derived from apparent wind speed, apparent wind angle and speed over ground.

Use the following Signal K dashboard menus:
- `Server/Plugin Config/Derived Data` in order to configure and start path derivation from existing paths; see https://github.com/SignalK/signalk-derived-data (you may have to instal the plugin first from Signal K `Appstore`)
- `Data Browser` in order to visualize the active and derived paths with their current values

#### Filtering by sources
Different sources may provide data for the same path; as needed, the `path` property may be extended to select a specific source, e.g. `navigation.speedOverGround[gps.1]`. The sources corresponding to a path on your Signal K server are identified by the property `$source` and can be obtained from the launch menu provided that the corresponding instruments are on and connected. A certain path may be listed several times in the chart specs, e.g.:
```javascript
{ path: "navigation.speedOverGround[gps.1]", AVG: "SOG1" },
{ path: "navigation.speedOverGround[gps.2]", AVG: "SOG2" },
{ path: "navigation.speedOverGround", AVG: "SOGx" },
```
The first two "path lines" will collect data each respectively from `gps.1` and from `gps.2`. The third line will collect SOG data from all sources (including the sources identified in line 1 and 2). 

An example is provided in [sources_filtering_example.js](./specs/sources_filtering_example.js)  (read the comments in the file before running: it runs on test log files provided with Signal K node server).

Note: Filtering by source relies on the property `$source` being present in the delta messages, which appears to be systematic with the current node.js implementation of Signalk. However the use of `$source` is not mandatory in Signalk specifications (it is then replaced by the object `source`). Hence, this feature may not work properly on other Signalk implementations.

#### Legends and lines colors
Colors can be specified per legend at the bottom of the chart specifications file in order to insure consistency accross multiple charts of a same set. Specification file [sail.js](./specs/sail.js) shows how to assign colors.

If not provided, colors will be assigned automatically by the c3 library.

### Unit conversion
[units.js](./specs/units.js) provides the list of Signal K units and charting units (not yet a complete set), with the conversion factors and algorithms.
It also provides the following default properties for the y and y2 axis as a function of the charting unit:
- `label`
- range (`min` and `max`)
- `tick.count` and `tick.format`

Those can be overridden in the chart specs.

A special unit `Percent` is provided. It allows to plot values of different units on a same "Percent" y or y2 axis by providing reference values in the Signal K unit for 0% and for 100%. See [engines.js](./specs/engines.js) for an example with comments. The same example shows how grid lines may be added to the chart. This may be used to mark normal and alert levels.

### General options
Options governing all charts are given in [stripcharts_options.js](./specs/stripcharts_options.js).
They are fully documented in the code.
Some comments also explain how time is managed in signalk-stripcharts.

A few options deal with the usage with InfluxDB: ignore those for the moment.

The options can be overidden by setting them in the chart specs files, e.g. `options.pointRadius = 2`.

Any of the following options can also be entered as query parameters after the url when launching stripcharts.html:
`useInfluxDB`, `timeTolSec`, `logTypes`, `policy`, `period` and `minPeriod`.

Options values are assigned in this order:
- from [stripcharts_options.js](./specs/stripcharts_options.js),
- then from the chart specs file,
- then from the url query parameter.

The last encountered value will prevail. 

Invalid option values will be replaced by defaults.

<div align="right"><sup><a  style="color:red;" href="#table-of-contents">↥ back to TOC</a></sup></div>

## Charting past data from InfluxDB

### General description

[InfluxDB](https://www.influxdata.com/products/influxdb-overview/) is an open source time series database designed to handle high write and query loads and provides a SQL-like query language called InfluxQL.
We will be using [InfluxDB 1.x](https://www.influxdata.com/time-series-platform/).

When the InfluxDB interface is not activated, signalk-stripcharts operates in the "**live mode**": the x-axis time origin is always the current time (possibly temporarily "paused").
With the InfluxDB interface, an additional mode is provided: the "**past mode**" which allows to choose another x-axis time origin in the past.

With the InfluxDB interface, the following behaviours are provided with two additional buttons:
- in **live mode**: while live data is being plotted and does not fill the complete time window yet, you can fill the right part of the chart with past data from an InfluxDB database by clicking on a "fill-right" button <sub>![fill-right button](./PNG/readme8.PNG)</sub>
- in **past mode**: after clicking on the "clock" button ![clock button](./PNG/readme9.PNG) , a pop-up calendar can be used to define a new origin for the x-axis and the chart will be filled with historical data from the InfluxDB database starting at and after that new time origin.

In order to use those features, you first need to:
- install InfluxDB and start it,
- create an InfluxDB database and configure it,
- provide the database name in [stripcharts_options.js](./specs/stripcharts_options.js),
- install and run the [Signal K to InfluxDB plugin](https://github.com/tkurki/signalk-to-influxdb) (this will write Signal K data to the database),

Detailed steps are  provided in the following sections. 
You may also refer to InfluxDB [get-started documentation](https://docs.influxdata.com/influxdb/v1.8/introduction/get-started/).

### Install InfluxDB and create a database

In order to install on a Raspberry PI refer to this [startup guide](https://pimylifeup.com/raspberry-pi-influxdb/).

Start the influx command line tool in a terminal:
```
influx
```
Then create the database "boatdata":
```
CREATE DATABASE boatdata
```


### Signalk to InfluxDB Plugin

[Signal K to InfluxDB plugin](https://github.com/tkurki/signalk-to-influxdb) can write all or selected simple numeric Signal K values to InfluxDB with a chosen sampling period (parameter `Resolution`).

From Signal K Dashboard select Server/Plugin Config > InfluxDB writer.
Fill the plugin form, e.g.:
- check `Active`
- change Host and Port as might be needed
- Database: `boatdata`
- Batch writes interval: `5` (or any other sensible value)
- Resolution (ms): `1000`
- If your are simulating from log files, consider checking `Override time with current timestamp`
- Define type of list and paths included/excluded
- Submit

### Provide the influxDB database in the options

The relevant lines in [stripcharts_options.js](./specs/stripcharts_options.js) are:
```
var options = {
  ...
  useInfluxDB: true,   // <===== YOU MUST SET TO TRUE as false is the default
  influxDB: {
    dbName: "boatdata",   // if missing 'boatdata' is the default
    SK2InfluxResolution_ms: 1000   // optional, default 1000
  }
  ...
}
```

Alternatively you could leave stripcharts_options.js unchanged and add the line `options.useInfluxDB = true;` at the top of the charts specs files for which you want InfluxDB access, or provide an url query parameter `...&useInfluxDB=true` when invoking the chart specs.

Note: the InfluxDB plugin `resolution` parameter value should be provided as SK2InfluxResolution_ms if not equal to 1000 msec. 

Now start a stripchart as explained above in [Basic usage](#basic-usage) and access past data as explained below.

An example is provided in [influx_test.js](./specs/influx_test.js)  (read the comments in the file before running: it runs on a test log file provided with Signal K node server)

Note: If you have been using InfluxDB before with several Retention Policies and Continuous Queries, you may need to further refine the options: read [Stripcharts options for InfluxDB](#stripcharts-options-for-influxdb)


### Usage in live mode and past mode

When `options.useInfluxDB` is `true` two buttons with **red icons** are added to the user interface. The window top line now looks like this:

![top line buttons](./PNG/readme7.PNG)

A "fill-right" button is added to the right of the y-axis buttons and a "clock button" is shown next to the top charts selection drop-down list.

When the charts are first displayed, the pause/play button has a red border: this indicates the "**live mode**" is active: the charts start being filled with live data. If you click the fill-right button <sub>![fill-right button](./PNG/readme8.PNG)</sub> the corresponding chart will be completely filled up to the right with past data from InfluxDB (if such data exist in the database). This action can be performed individually for each chart. Once a chart is filled up the chart fill-right button is hidden.

Let's now explore and chart older data.

Click the clock-button ![clock button](./PNG/readme9.PNG) . A date/time picker window pops up (with the current GMT time). The clock-button is now red-bordered indicating that the "**past mode**" is active (while in that mode and unless Signal K is disconnected, Signal K live data continue to be received and buffered "behind the scene").

Pick a date/time a few minutes, hours or days in the past and click `OK` : the origins of both charts in view are set to the picked time and the charts are filled with historical data from InfluxDB provided that InfluxDB was active at that time and contains historical data.

You can further explore the past by selecting another chart in the drop-down list or by clicking the clock-button and selecting another date/time.

When the play button is clicked, the **live mode** is again activated, the charts x-axis are set to the current time and, unless SignalK was disconnected, become live-updated again (playing state).


### Optimize with retention periods and constant queries

You can use the database created as is, however there is a way to configure it in order to optimize disk space and query resources consumption while providing access to a long history of boat data.

A natural solution is to downsample the data, i.e. keep the high precision raw data for a limited time, and store the lower precision, summarized data longer, as explained here: [downsampling and data retention](https://docs.influxdata.com/influxdb/v1.8/guides/downsample_and_retain/). It relies on the concepts of Retention Policy (RP) and Continuous Query (CQ). 

The following target configuration will be configured, chosen to be compatible with the sampling periods (avgInterval) of the charts specs provided with signalk-stripcharts:

| RPs & CQs          | "1day"  |   ---"CQ1"-->    | "1week"|    ---"CQ2"-->    | "1month"|
| -------------------| :------:| :---------------:| :-----:| :----------------:| :------:|

CHANGELOG.md


## Please see [Releases](https://github.com/edefalque/signalk-stripcharts/releases)  for the release notes.


sail.js
"use strict";
// specifications for several concurrent charts (2 of them displayed at once)
// Note: all durations in sec

//=================== Wind speeds 10min, 2h, 24h =======================

const Wind_speeds_10min = 
    // for javascript variables naming rules see https://javascript.info/variables#variable-naming
    { stripChartName: "Wind_speeds_10min",   // use preferably the same as the containing object name
        timeWindow: 10*60,          // 10min = 10 * 60
        avgInterval: 2,             // 2 sec
        intervalsPerRefresh: 2,     // default 1
        x: { label: "min/sec ago", tickCount: 11 },   // a tick every min (10, + 1 for zero)
        y: { unit: "Knot" },
        paths: 
        [
            { path: "environment.wind.speedTrue", AVG: "TWS" },
            { path: "environment.wind.speedApparent", AVG: "AWS" }      
        ]
    }
;

const Wind_speeds_2h = derivedFrom(Wind_speeds_10min,  // inheritance
    { stripChartName: "Wind_speeds_2h",
        timeWindow: 2*60*60,       // 2h
        avgInterval: 10,           // 10 sec
        intervalsPerRefresh: 1,    // overide inherited value 2 !
        x: { label: "hrs/min ago", tickCount: 13 }  // a tick every 10 min (12, + 1 for zero)
    }
);

const Wind_speeds_24h = derivedFrom(Wind_speeds_2h,    // inheritance
    { stripChartName: "Wind_speeds_24h",
    skip: false,
        timeWindow: 24*60*60,      // 24h
        avgInterval: 120,          // 120 sec
        paths: 
        [
            { path: "environment.wind.speedTrue", MAX: "maxTWS", MIN: "minTWS" },
            { path: "environment.wind.speedApparent",  MAX: "maxAWS", MIN: "minAWS" }      
        ]
    }
);

//=================== Wind angles 10min, 2h, 24h ======================= 

const Wind_angles_10min =
    { stripChartName: "Wind_angles_10min",
        timeWindow: 600,          // 10min
        avgInterval: 2,           // 2 sec
        intervalsPerRefresh: 2,   // default 1
        x: { label: "min/sec ago", tickCount: 11 },   // a tick every min (10, + 1 for zero)
        y: { unit: "Angle_Degree" },
        paths: 
        [
            { path: "environment.wind.angleTrueWater", AVG: "TWA" },
            { path: "environment.wind.angleApparent", AVG: "AWA" },     
            { path: "performance.beatAngle", AVG: "polBA" },      
            { path: "performance.gybeAngle", AVG: "polGA" }      
        ]
    };

const Wind_angles_2h = derivedFrom(Wind_angles_10min,
    { stripChartName: "Wind_angles_2h",
        timeWindow: 2*60*60,      // 2h
        avgInterval: 10,          // 10 sec
        intervalsPerRefresh: 1,
        x: { label: "hrs/min ago", tickCount: 13 },
    }
);

const Wind_angles_24h = derivedFrom(Wind_angles_2h,
    { stripChartName: "Wind_angles_24h",
        timeWindow: 24*60*60,     // 24h
        avgInterval: 120,         // 120 sec
        paths: 
        [
            { path: "environment.wind.angleTrueWater", MAX: "maxTWA", MIN: "minTWA" },
            { path: "environment.wind.angleApparent", MAX: "maxAWA", MIN: "minAWA" }      
        ]
    }
);

//=================== Boat speeds 10min, 2h, 24h ======================= 

const Boat_speeds_10min =
    { stripChartName: "Boat_speeds_10min",
        timeWindow: 600,          // 10min
        avgInterval: 2,           // 2 sec
        intervalsPerRefresh: 2,   // default 1
        x: { label: "min/sec ago", tickCount: 11 },   // a tick every min (10, + 1 for zero)
        y: { unit: "Knot" },
        paths: 
        [
            { path: "navigation.speedOverGround", AVG: "SOG" },
            { path: "navigation.speedThroughWater", AVG: "STW" },      
            { path: "performance.velocityMadeGood", AVG: "VMG" },     
            { path: "performance.polarSpeed", AVG: "polS" } 
        ]
    };

const Boat_speeds_2h = derivedFrom(Boat_speeds_10min,
    { stripChartName: "Boat_speeds_2h",
        timeWindow: 2*60*60,      // 2h
        avgInterval: 10,          // 10 sec
        intervalsPerRefresh: 1,
        x: { label: "hrs/min ago", tickCount: 13 },
    }
);

const Boat_speeds_24h = derivedFrom(Boat_speeds_2h,
    { stripChartName: "Boat_speeds_24h",
        timeWindow: 24*60*60,     // 24h
        avgInterval: 120,         // 120 sec
        paths: 
        [
            { path: "navigation.speedOverGround", MAX: "maxSOG", MIN: "minSOG" },
            { path: "navigation.speedThroughWater", MAX: "maxSTW", MIN: "minSTW" }     
        ]
    }
);

//=================== Wind/boat directions 10min, 2h, 24h ======================= 

const Wind_boat_dir_10min =
    { stripChartName: "Wind_boat_dir_10min",
        timeWindow: 600,          // 10min
        avgInterval: 2,           // 2 sec
        intervalsPerRefresh: 2,   // default 1
        x: { label: "min/sec ago", tickCount: 11 },   // a tick every min (10, + 1 for zero)
        y: { label: "Degrees True", unit: "Direction_Degree" },
        paths: 
        [
            { path: "environment.wind.directionTrue", AVG: "TWD" },
            { path: "navigation.courseOverGroundTrue", AVG: "COG" },      
            { path: "navigation.headingTrue", AVG: "HDT" }      
        ]
    };

const Wind_boat_dir_2h = derivedFrom(Wind_boat_dir_10min,
    { stripChartName: "Wind_boat_dir_2h",
        timeWindow: 2*60*60,      // 2h
        avgInterval: 10,          // 10 sec
        intervalsPerRefresh: 1,
        x: { label: "hrs/min ago", tickCount: 13 },
    }
);

const Wind_boat_dir_24h = derivedFrom(Wind_boat_dir_2h,
    { stripChartName: "Wind_boat_dir_24h",
        timeWindow: 24*60*60,    // 24h
        avgInterval: 120,        // 120 sec
    }
);

// ===================== stripCharts set ============================
// define stripChartsSpecs as a set of selectable stripCharts
// identify the two charts that will display at startup (optional)

    const  stripChartsSpecs = {
        name: "Sail",    // shows in tab title
        stripCharts: [
            Wind_speeds_10min,
            Wind_speeds_2h,
            Wind_speeds_24h,
            Wind_angles_10min,
            Wind_angles_2h,
            Wind_angles_24h,
            Boat_speeds_10min,
            Boat_speeds_2h,
            Boat_speeds_24h,
            Wind_boat_dir_10min,
            Wind_boat_dir_2h,
            Wind_boat_dir_24h
        ],
        initialUpperView: Wind_speeds_10min,    // one of the stripCharts
        initialLowerView: Wind_angles_10min     // another one or "none"
};

// ===================== Color palette ============================
// Object with colors for legends, lines, etc.;
// method colorFor(legend) returns a string with the color code or name
// Note: if no colorPalette is provided, colors will be chosen by c3 library
const colorPalette = {
	// Picked from https://www.w3schools.com/colors/colors_picker.asp   (see "Hue")
	// Using hsl (hue, saturation, lightness) notation, fully saturated colors are selected
	// For AVG lines, lightness is 50%
	// MAX lines are lighter (approx +25%)
	// MIN lines are darker (approx -20%) 
	// Apparent wind speeds/angles are greenish
		AWS: 	  "hsl(120, 100%, 50%)",
		maxAWS: "hsl(120, 100%, 75%)",
		minAWS: "hsl(120, 100%, 30%)",
		AWA: 	  "hsl(120, 100%, 50%)",
		maxAWA: "hsl(120, 100%, 75%)",
		minAWA: "hsl(120, 100%, 30%)",
	// True wind speeds/angles are blueish
		TWS: 	  "hsl(215, 100%, 50%)", 
		maxTWS: "hsl(215, 100%, 75%)",
		minTWS: "hsl(215, 100%, 30%)",
		polBA: "hsl(215, 100%, 90%)",
		polGA: "hsl(215, 100%, 90%)",
		TWA: 	  "hsl(215, 100%, 50%)", 
		maxTWA: "hsl(215, 100%, 75%)",
		minTWA: "hsl(215, 100%, 30%)",
	// True Wind directions are Cyan-ish
		TWD: 	  "hsl(180, 100%, 50%)", 
		maxTWD: "hsl(180, 100%, 80%)",
		minTWD: "hsl(180, 100%, 30%)",
	// Boat water speeds/headings are redish
		STW: 	  "hsl(0, 100%, 50%)",
		maxSTW: "hsl(0, 100%, 75%)",
		minSTW: "hsl(0, 100%, 30%)",
		polS: "hsl(0, 100%, 90%)",
		HDT: 	  "hsl(0, 100%, 50%)",
		maxHDT: "hsl(0, 100%, 75%)",
		minHDT: "hsl(0, 100%, 30%)",
	// Boat speeds/directions Over Ground are purple-ish
		SOG: 	  "hsl(300, 100%, 50%)",
		maxSOG: "hsl(300, 100%, 75%)",
		minSOG: "hsl(300, 100%, 30%)",
		COG: 	  "hsl(300, 100%, 50%)",
		maxCOG: "hsl(300, 100%, 75%)",
		minCOG: "hsl(300, 100%, 30%)",
	// VMG is goldish
		VMG: 	  "hsl(51, 100%, 50%)",
		maxVMG: "hsl(51, 100%, 75%)",
		minVMG: "hsl(51, 100%, 30%)",
	// depths redish
		DEPTH:    "hsl(0, 100%, 50%)",
		maxDEPTH: "hsl(0, 100%, 75%)",
		minDEPTH: "hsl(0, 100%, 35%)",
	// Sea water temperatures brownish
		WTEMP:    "hsl(25, 100%, 40%)",
		maxWTEMP: "hsl(25, 100%, 55%)",
		minWTEMP: "hsl(25, 100%, 25%)",

	colorFor: function(legend) {
		if (this.hasOwnProperty(legend)) {
			return this[legend];
		}
		else  {
			console.log( "no color in palette for: " + legend);
			return "Grey";	// Note: would an empty string be returned, c3 would choose a color for you
		}
	}
}
